subprojects/core/src/main/java/org/gradle/caching/internal/tasks/DefaultTaskOutputCachingBuildCacheKeyBuilder.java

```diff
@@ -90,11 +97,18 @@ public void appendTaskActionImplementations(Collection<ImplementationSnapshot> t
     }
 
     @Override
-    public void appendInputPropertyHash(String propertyName, HashCode hashCode) {
+    public void appendInputPropertyHash(String propertyName, @Nullable HashCode hashCode) {
         hasher.putString(propertyName);
-        hasher.putHash(hashCode);
-        inputHashes.put(propertyName, hashCode);
-        LOGGER.info("Appending inputPropertyHash for '{}' to build cache key: {}", propertyName, hashCode);
+        if (hashCode != null) {
+            hasher.putHash(hashCode);
+            inputHashes.put(propertyName, hashCode);
+            LOGGER.info("Appending inputPropertyHash for '{}' to build cache key: {}", propertyName, hashCode);
+        } else {
+            valid = false;
+            // Add a random hash to make sure it shows up as changed in the build scan
+            inputHashes.put(propertyName, HashCode.fromInt(new Random().nextInt()));
+            LOGGER.info("Could not append invalid inputPropertyHash for '{}' to build cache key.", propertyName);
+        }
     }
```

Comment:
Again, let's not use `null` here, but instead handle this case explicitly. As a side-effect it would allow us to print some more meaningful message, too, like "The implementation of this thing cannot be determined, because it was loaded via some unknown classloader".

Reply:
I use a separate method now.